syntax = "proto3";

package envoy.type;

import "validate/validate.proto";
import "gogoproto/gogo.proto";

option (gogoproto.equal_all) = true;

// [#protodoc-title: Percent]

// Identifies a percentage, in the range [0.0, 100.0].
message Percent {
  double value = 1 [(validate.rules).double = {gte: 0, lte: 100}];
}

// A fractional percentage is used in cases in which for performance reasons performing floating
// point to integer conversions during randomness calculations is undesirable. The message includes
// both a numerator and denominator that together determine the final fractional value.
//
// * **Example**: 1/100 = 1%.
// * **Example**: 3/10000 = 0.03%.
message FractionalPercent {
  // Specifies the numerator. Defaults to 0.
  uint32 numerator = 1;

  // Fraction percentages support several fixed denominator values.
  enum DenominatorType {
    // 100.
    //
    // **Example**: 1/100 = 1%.
    HUNDRED = 0;

    // 10,000.
    //
    // **Example**: 1/10000 = 0.01%.
    TEN_THOUSAND = 1;

    // 1,000,000.
    //
    // **Example**: 1/1000000 = 0.0001%.
    MILLION = 2;
  }

  // Specifies the denominator. If the denominator specified is less than the numerator, the final
  // fractional percentage is capped at 1 (100%).
  DenominatorType denominator = 2 [(validate.rules).enum.defined_only = true];
}

// Runtime derived FractionalPercent with defaults for when the numerator or denominator is not
// specified via a runtime key.
message RuntimeFractionalPercent {
  // Default value if the runtime value's for the numerator/denominator keys are not available.
  FractionalPercent default_value = 1 [(validate.rules).message.required = true];

  // Runtime key to get numerator value for comparison. This value is used if defined.
  string numerator_runtime_key = 2;

  // Runtime key to get denominator type for comparison. This value is used if defined. Valid field
  // value strings must match valid :ref:`denominator
  // types<envoy_api_enum_type.FractionalPercent.DenominatorType>`.
  string denominator_runtime_key = 3;
}
